File: 3203. Find Minimum Diameter After Merging Two Trees/main.go
package main

import "math"

func buildAdj(edges [][]int) map[int][]int {
	adj := make(map[int][]int)
	for _, edge := range edges {
		n1, n2 := edge[0], edge[1]
		adj[n1] = append(adj[n1], n2)
		adj[n2] = append(adj[n2], n1)
	}
	return adj
}

func getDiameter(cur, par int, adj map[int][]int) (int, int) {
	maxD := 0
	firstMax, secondMax := 0, 0

	for _, nei := range adj[cur] {
		if nei == par {
			continue
		}
		neiD, neiMaxLeafPath := getDiameter(nei, cur, adj)
		maxD = int(math.Max(float64(maxD), float64(neiD)))

		if neiMaxLeafPath > firstMax {
			secondMax = firstMax
			firstMax = neiMaxLeafPath
		} else if neiMaxLeafPath > secondMax {
			secondMax = neiMaxLeafPath
		}
	}

	maxD = int(math.Max(float64(maxD), float64(firstMax+secondMax)))
	return maxD, 1 + firstMax
}

func minimumDiameterAfterMerge(edges1, edges2 [][]int) int {
	adj1 := buildAdj(edges1)
	adj2 := buildAdj(edges2)

	d1, _ := getDiameter(0, -1, adj1)
	d2, _ := getDiameter(0, -1, adj2)

	return int(math.Max(float64(d1), math.Max(float64(d2), math.Ceil(float64(d1)/2)+math.Ceil(float64(d2)/2)+1)))
}

File: 3203. Find Minimum Diameter After Merging Two Trees/main.py
from collections import defaultdict
import heapq
from math import ceil
from typing import List


class Solution:
    def minimumDiameterAfterMerge(self, edges1: List[List[int]], edges2: List[List[int]]) -> int:
        def build_adj(edges):
            adj = defaultdict(list)
            for n1, n2 in edges:
                adj[n1].append(n2)
                adj[n2].append(n1)
            return adj
        
        adj1 = build_adj(edges1)
        adj2 = build_adj(edges2)

        # DFS, return [diam, max_leaf_path]
        def get_diameter(cur, par, adj):
            max_d = 0
            max_child_paths = [0, 0]

            for nei in adj[cur]:
                if nei == par: continue
                nei_d, nei_max_leaf_path = get_diameter(nei, cur, adj)
                max_d = max(max_d, nei_d)
                heapq.heappush(max_child_paths, nei_max_leaf_path)
                heapq.heappop(max_child_paths)
            
            max_d = max(max_d, sum(max_child_paths))
            return [max_d, 1 + max(max_child_paths)]

        d1, _ = get_diameter(0, -1, adj1)
        d2, _ = get_diameter(0, -1, adj2)

        return max(d1, d2, ceil(d1/2) + ceil(d2/2) + 1)
